packages/ens-utils/src/registrar.ts

```diff
@@ -0,0 +1,217 @@
+
+  /**
+   * The action associated with this `RegistrarPriceQuote`.
+   */
+  action: RegistrarAction;
+
+  /**
+   * The `TimePeriod` that this `RegistrarPriceQuote` is for.
+   */
+  rentalPeriod: TimePeriod;
+
+  /**
+   * The set of distinct `RegistrarCharge` values that must be paid to the
+   * `Registrar` to rent the domain for the period of `rentalPeriod`.
+   * 
+   * May be empty if the domain is free to rent for `rentalPeriod`.
+   */
+  charges: RegistrarCharge[];
+
+  /**
+   * The total price to rent the domain for the period of `rentalPeriod`.
+   * 
+   * This is the sum of `charges` in the unit of currency that should be paid
+   * to the registrar.
+   */
+  totalPrice: Price;
+}
+
+export interface RegistrationPriceQuote extends AbstractRegistrarPriceQuote {
+  action: typeof RegistrarAction.Register;
+}
+
+export interface RenewalPriceQuote extends AbstractRegistrarPriceQuote {
+  action: typeof RegistrarAction.Renew;
+}
+
+export interface Registrar {
+  getManagedSubname(name: ENSName): ENSName | null;
+
+  // NOTE: Throws RegistrarUnsupportedNameError if name is not supported by this registrar
+  getValidatedSubname(name: ENSName): ENSName;
+
+  getOnchainRegistrar(): ContractRef | null;
+
+  canRegister(name: ENSName, atTimestamp: Timestamp, duration: Duration, existingRegistration?: Registration): boolean;
+  getRegistrationPriceQuote(name: ENSName, atTimestamp: Timestamp, duration: Duration, existingRegistration?: Registration): RegistrationPriceQuote;
+  
+  canRenew(name: ENSName, atTimestamp: Timestamp, duration: Duration, existingRegistration: Registration): boolean;
+  getRenewalPriceQuote(name: ENSName, atTimestamp: Timestamp, duration: Duration, existingRegistration: Registration): RenewalPriceQuote;
+}
+
+export interface OnchainRegistrar extends Registrar {
+  getOnchainRegistrar(): ContractRef;
+}
+
+export interface OffchainRegistrar extends Registrar {
+  getOnchainRegistrar(): null;
+}
+
+/**
+ * Identifies that a name was passed to a `Registrar` function that was not relevant for that `Registrar`.
+ * 
+ * @param message a reason why `name` was NOT relevant for `Registrar`.
+ * @param label the `ENSName` value that was NOT relevant.
+ */
+export class RegistrarUnsupportedNameError extends Error {
+
+  public constructor(message: string, name: ENSName) {
+    super(`RegistrarUnsupportedNameError for name "${name.name}": ${message}`);
+    this.name = "RegistrarUnsupportedNameError";
+    Error.captureStackTrace(this, this.constructor);
+  }
+}
+
+export const scaleAnnualPrice = (annualPrice: Price, duration: Duration) => {
+  // Small performance optimization if no scaling is needed
+  if (duration.seconds === SECONDS_PER_YEAR.seconds)
+    return annualPrice;
+
+  // TODO: verify we're doing this division of bigints correctly
+  const scaledRate = Number(duration.seconds) / Number(SECONDS_PER_YEAR.seconds);
+
+  return approxScalePrice(annualPrice, scaledRate);
+}
```

Comment:
I'd rather do it this way:

```suggestion
export const scaleAnnualPrice = (annualPrice: Price, duration: Duration) => {
  // Small performance optimization if no scaling is needed
  if (duration.seconds === SECONDS_PER_YEAR.seconds)
    return annualPrice;

  const scaledRate = (duration.seconds) / (SECONDS_PER_YEAR.seconds);

  return approxScalePrice(annualPrice, Number(scaledRate));
}
```

Since a rate shall be smaller than a seconds value, making the seconds representation as bigint and the rate as number.

BUT

As divisions by bigint will never result in floating point numbers, for precision sake and with the understandment that, hardly some seconds value will be greater than `Number.MAX_SAFE_INTEGER` (9007199254740991 which represents date that is later than 259000)

Reply:
Considering above information, you are doing the division correctly and safely

Reply:
Thanks do you still have a suggestion that we change anything here?

Reply:
Appreciate if you could please create unit tests for this function. That's the best way to be certain.